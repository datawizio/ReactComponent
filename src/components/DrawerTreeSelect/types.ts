import { SelectValue } from "antd/lib/tree-select";

import { TreeSelectProps as AntTreeSelectProps } from "./antd/AntTreeSelect";

import { AntTreeNode } from "antd/lib/tree";

export interface IDrawerTreeSelectFilters {
  shop_markers?: string[];
  search?: string;
  level?: string | number;
  value?: SelectValue;
  first?: boolean;
}

export type LevelsType = { value: string; label: string }[];

export interface MarkersRenderProps {
  onChange?: (selected: string[]) => void;
}

export interface DrawerTreeSelectProps<VT>
  extends Omit<AntTreeSelectProps<VT>, "onChange" | "loadData"> {
  additionalFilters?: any;
  asyncData?: boolean;

  headerHeight?: number;

  /**
   * Title Drawer-а
   */
  drawerTitle?: string;

  /**
   * Drawer width in px
   */
  drawerWidth?: number;

  /**
   * Показать выбор левелов
   */
  showLevels?: boolean;

  levels?: LevelsType;

  level?: string | number;

  /**
   * tree data is flat list or not
   */
  isFlatList?: boolean;

  remoteSearch?: boolean;

  showSelectAll?: boolean;

  selectAllText?: string;

  treeDataCount?: number;

  emptyIsAll?: boolean;

  selectedMarkers?: string[];

  markersRender?: ((props: MarkersRenderProps) => React.ReactElement) | null;

  loadData?: (filters: IDrawerTreeSelectFilters) => Promise<any>;

  loadChildren?: (nodeId: string, filters?: any) => Promise<any>;
  /**
   * Event when user click Submit
   */
  onChange?: (values: SelectValue, selected?: AntTreeNode) => void;
  onChangeReturnObject?: (obj: {
    value: SelectValue;
    level: string | number;
    selected?: AntTreeNode;
  }) => void;

  onLevelChange?: (level: string) => void;
}

export interface FCDrawerTreeSelect<VT>
  extends React.FC<DrawerTreeSelectProps<VT>> {
  SHOW_PARENT: "SHOW_PARENT";
  SHOW_ALL: "SHOW_ALL";
  SHOW_CHILD: "SHOW_CHILD";
}
