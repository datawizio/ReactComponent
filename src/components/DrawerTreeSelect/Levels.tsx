import React from "react";

import Select from "../Select";

interface LevelsProps {
  value?: number;
  levels?: { value: string; label: string }[];
  onChange?: (value: number) => void;
}

export const Levels: React.FC<LevelsProps> = ({ value, levels, onChange }) => {
  const handleChange = value => {
    onChange(value);
  };

  return (
    <Select
      defaultValue={value}
      options={levels}
      onChange={handleChange}
      className="drawer-tree-select-levels"
    />
  );
};

Levels.defaultProps = {
  value: 1,
  levels: []
};
