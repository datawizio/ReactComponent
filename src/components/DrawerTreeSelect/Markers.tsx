import React, { useCallback } from "react";
import { useTranslation } from "react-i18next";
import AntTreeSelect from "./antd/AntTreeSelect";
import TreeSelect from "../TreeSelect";

interface MarkersProps {
  value?: string[] | number[];
  treeData?: any;
  onChange?: any;
  loadChildren?: (id: string, filters?: any) => Promise<any>;
  placeholder?: string;
  style?: React.CSSProperties;
}

export const Markers: React.FC<MarkersProps> = React.memo(
  ({ value, treeData, onChange, loadChildren, placeholder, style }) => {
    const { t } = useTranslation();

    const handleChange = value => {
      onChange && onChange(value);
    };

    const handleSearch = useCallback((inputValue: string, treeNode: any) => {
      return treeNode.title.toLowerCase().includes(inputValue.toLowerCase());
    }, []);

    const loadData = async node => {
      if (node.children?.length) return;
      await loadChildren(node);
    };

    return (
      <TreeSelect
        value={value}
        treeData={treeData}
        treeCheckable={true}
        placeholder={placeholder ?? t("SHOP_MARKERS")}
        showCheckedStrategy={AntTreeSelect.SHOW_PARENT}
        loadData={loadData}
        onChange={handleChange}
        filterTreeNode={handleSearch}
        className="select-markers-field"
        style={style}
      />
    );
  }
);

Markers.defaultProps = {
  value: [],
  treeData: []
};
