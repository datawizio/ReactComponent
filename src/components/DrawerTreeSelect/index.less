@import "~antd/es/select/style/index";
@import "~antd/es/radio/style/index";
@import "~antd/es/tree-select/style/index";
@import "../../styles/variables";
@import "./rtl";

.drawer-tree-select.ant-select-focused .ant-select-selection-search-input,
.drawer-tree-select-dropdown-fake.ant-select-dropdown.ant-tree-select-dropdown {
  visibility: hidden;
}

.drawer-tree-select .ant-select-selection-search-input {
  font-size: 1px;
}

.drawer-tree-select {
  .ant-select-selection-item {
    .ant-select-selection-item-content {
      border: 0;
      background: transparent;
      margin: 0;
      padding: 0;
      font-size: 14px;

      .anticon-close {
        font-size: 12px;
      }
    }
  }

  span {
    & > .ant-select-selection-placeholder {
      display: none;
    }

    &:first-of-type > .ant-select-selection-placeholder {
      display: block;
    }
  }

  &-loading-placeholder {
    background: @ant-tag-background-color;
    border: 1px solid @ant-tag-border-color;
    padding-left: 5px;
    border-radius: 0 3px 3px 0;
    color: @ant-tag-color;
  }
}

.drawer-tree-selected-all {
  .ant-select-selection-overflow-item-rest {
    .ant-select-selection-item {
      display: none;
    }
  }
  .ant-select-selection-placeholder {
    color: @select-multiple-item-disabled-color;
    font-weight: 300;
  }
}

.drawer-tree-select-dropdown {
  .ant-drawer-body {
    overflow: hidden;
    padding-bottom: @padding-lg - 16px;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;

    div + .ant-input-search,
    div + .drawer-tree-select-levels {
      margin-top: @margin-sm;
    }

    .ant-input-search {
      margin-bottom: 10px;
    }

    .ant-select-empty {
      margin-top: @margin-sm;
    }

    .drawer-tree-select-levels {
      width: 100%;
      display: flex;

      label {
        flex: 1;
        text-align: center;
        padding: 0;
      }
    }

    .drawer-tree-select-dropdown-toolbar {
      margin-bottom: 4px;
      min-height: 24px;

      .ant-checkbox + span {
        padding-left: @padding-sm;
      }
    }

    .drawer-tree-select-selected {
      padding-top: 5px;
      display: flex;
      color: @drawer-selected-text-color;
      background: inherit;

      .drawer-tree-select-selected-title {
        flex: 1;
      }
    }

    .ant-select-tree {
      .tree-check-all {
        display: none;
      }

      // padding-top: @padding-md;
      margin-left: -@margin-lg;

      .tree-check-all .ant-select-tree-switcher {
        visibility: hidden;
      }
    }

    .drawer-select-loader-container {
      height: 16px;
      flex: 1;
      overflow: visible;

      .ant-skeleton-title {
        margin: 0;
      }

      .ant-skeleton-content .ant-skeleton-title + .ant-skeleton-paragraph {
        margin-top: 5px;

        li {
          width: 300px !important;
        }
      }
    }
  }

  &.drawer-tree-select-show-all {
    .ant-drawer-body {
      .ant-select-tree {
        .tree-check-all {
          display: flex;
        }
      }
    }
  }

  &.drawer-tree-select-dropdown-flat-list {
    .ant-drawer-body {
      .ant-select-tree {
        .ant-select-tree-indent {
          display: none;
        }
      }
    }
  }
}