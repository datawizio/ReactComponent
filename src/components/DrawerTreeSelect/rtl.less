.drawer-tree-select-dropdown {
  .ant-drawer-body .ant-select-tree {
    margin-left: unset;
    margin-right: -24px;
  }
}