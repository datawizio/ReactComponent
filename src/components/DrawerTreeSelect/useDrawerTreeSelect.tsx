import { useReducer } from "react";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { SelectValue } from "../DrawerSelect/antd/AntSelect";
import { AntTreeNode } from "antd/es/tree";
import { LevelsType } from "./types";
import { Key } from "antd/es/table/interface";

export interface IUseDrawerTreeSelect {
  drawerVisible: boolean;
  internalValue: any;
  selected: AntTreeNode;
  stateTreeData: any;
  internalLoading: boolean;
  internalLevels: LevelsType;
  selectAllState: string;
  internalTreeDataCount: number;
  internalTreeExpandedKeys: Key[];
  fakeVisible: boolean;
}

function reducer(state: IUseDrawerTreeSelect, action: any) {
  switch (action.type) {
    case "resetInternalValue":
      return { ...state, internalValue: [] };
    case "remoteLoadDataStart": {
      const newState: any = {
        stateTreeData: [],
        internalLoading: true
      };
      if (action.payload) {
        newState.value = action.payload;
      }
      return { ...state, ...newState };
    }
    case "remoteLoadDataStop": {
      action.payload.internalValue = filterInternalValue(state, action.payload.stateTreeData);
      return {
        ...state,
        internalLoading: false,
        ...action.payload
      };
    }
    case "drawerCancel": {
      action.payload.internalValue = filterInternalValue(state, action.payload.stateTreeData);
      return {
        ...state,
        selectAllState: "",
        drawerVisible: false,
        fakeVisible: false,
        internalTreeExpandedKeys: [],
        ...action.payload
      };
    }
    case "openDrawer": {
      return {
        ...state,
        drawerVisible: true,
        fakeVisible: true,
        ...action.payload
      };
    }
    case "drawerSubmit": {
      return {
        ...state,
        drawerVisible: false,
        fakeVisible: false,
        internalTreeExpandedKeys: []
      };
    }
    case "setSelected": {
      return {
        ...state,
        selected: action.payload
      };
    }
    case "internalTreeExpandedKeys": {
      return {
        ...state,
        internalTreeExpandedKeys: action.payload
      };
    }
    case "stateTreeData": {
      return {
        ...state,
        stateTreeData: action.payload
      };
    }
    case "internalValue": {
      return {
        ...state,
        internalValue: action.payload
      };
    }
    case "setState": {
      return {
        ...state,
        ...action.payload
      };
    }
    default:
      throw new Error();
  }
}

export const filterInternalValue = (state, stateTreeData) => {
  // check if selected keys are present in tree
  // filter internalValue by payload stateTreeData
  const stateTreeDataSet = new Set(stateTreeData.map(item => item.value));
  if (state.internalValue.length) {
    return state.internalValue.filter(item => {
      return stateTreeDataSet.has(item);
    });
  }
  return [];
}

export const useDrawerTreeSelect = (initialState: IUseDrawerTreeSelect) => {
  const [state, dispatch] = useReducer<any>(reducer, initialState);

  return [state as IUseDrawerTreeSelect, dispatch as any];
};
