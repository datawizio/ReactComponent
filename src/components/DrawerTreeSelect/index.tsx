import React, {
  useState,
  useRef,
  useCallback,
  useMemo,
  useEffect
} from "react";
import clsx from "clsx";

import { Skeleton, Tag } from "antd";
import SearchInput from "../SearchInput";
import Drawer from "../Drawer";
import Button from "../Button";
import { Levels } from "./Levels";
import AntTreeSelect from "./antd/AntTreeSelect";
import Checkbox from "../Checkbox";

import { triggerInputChangeValue } from "../../utils/trigger";

import { IDrawerTreeSelectFilters, FCDrawerTreeSelect } from "./types";
import { SelectValue } from "antd/lib/tree-select";
import { DataNode } from "rc-tree-select/es/interface";

import "./index.less";
import { useDrawerTreeSelect } from "./useDrawerTreeSelect";

function getMainLevelItems(items: any[], level: string | number | null = 1) {
  const set = new Set<string>();

  for (let item of items) {
    if ((level === null && item.pId === null) || item.level === level) {
      set.add(item.id);
    } else {
      break;
    }
  }

  return set;
}

function isAllItemsChecked(items: string[], set: Set<string>) {
  if (items.length !== set.size) return false;

  for (const value of items) {
    if (!set.has(value)) {
      return false;
    }
  }

  return true;
}
function calcEmptyIsAll(
  emptyIsAll: boolean,
  filters: IDrawerTreeSelectFilters
) {
  if (!emptyIsAll) return false;
  if (filters.search) return false;
  if (filters.level.toString() !== "1") return false;
  if (filters.formats && filters.formats.length !== 0) return false;
  return true;
}

const DrawerTreeSelect: FCDrawerTreeSelect<SelectValue> = ({
  additionalFilters,
  asyncData,
  showLevels,
  levels,
  level,
  drawerTitle,
  drawerSearchPlaceholder,
  drawerWidth,
  formatRender,
  treeDefaultExpandedKeys,
  treeExpandedKeys,
  treeData,
  treeDataCount,
  cancelText,
  submitText,
  loadingText,
  noDataText,
  levelText,
  value,
  isFlatList,
  onChange,
  onLevelChange,
  loadData,
  loadChildren,
  showCheckedStrategy,
  multiple,
  remoteSearch,
  loading,
  showSelectAll,
  selectAllText,
  emptyIsAll,
  placeholder,
  ...restProps
}) => {
  const [
    {
      drawerVisible,
      internalValue,
      selected,
      stateTreeData,
      internalLoading,
      internalLevels,
      selectAllState,
      internalTreeDataCount,
      internalTreeExpandedKeys
    },
    dispatch
  ] = useDrawerTreeSelect({
    drawerVisible: false,
    internalValue: value,
    selected: undefined,
    stateTreeData: treeData,
    internalLoading: loading,
    internalLevels: levels,
    selectAllState: "",
    internalTreeDataCount: 0,
    internalTreeExpandedKeys: []
  });

  const [, setSearchValue] = useState<string>("");
  const mainLevelItems = useRef<Set<string>>();

  const formatSelected = useRef<string[]>([]);
  const searchValue = useRef<string>();
  const levelSelected = useRef<string | number | null>(
    showLevels ? level : null
  );

  const prevTreeData = useRef<DataNode[]>();
  const prevLevel = useRef<string | number>();
  const prevFormatSelected = useRef<string[]>();
  const prevEmptyIsAllRef = useRef<boolean>();

  const drawerVisibleRef = useRef<boolean>(false);

  const emptyIsAllRef = useRef<boolean>(emptyIsAll);

  const inputRef = useRef<HTMLInputElement>();

  const showAllRef = useRef<boolean>(false);

  const internalTreeDefaultExpandedKeys = useMemo(() => {
    if (searchValue.current && !remoteSearch) return undefined;
    if (internalTreeExpandedKeys.length > 0) return internalTreeExpandedKeys;

    return treeDefaultExpandedKeys;
  }, [
    treeDefaultExpandedKeys,
    remoteSearch,
    searchValue,
    internalTreeExpandedKeys
  ]);

  const isLevelShowed =
    showLevels && internalLevels && internalLevels.length > 1;

  // ----- METHODS -------

  const setInputRef = (el: HTMLInputElement) => {
    if (el.classList.contains("ant-select-selection-search-input")) {
      inputRef.current = el;
    }
  };

  const triggerOnChange = useCallback(
    value => {
      if (!onChange) return;
      if (!multiple) {
        if (value[0]) return onChange(value[0], selected);
        return onChange("");
      }
      if (isSelectedAll) {
        dispatch({ type: "resetIntervalValue" });
        return onChange([]);
      }
      onChange(value);
    },
    //eslint-disable-next-line
    [multiple, selected, onChange, selectAllState, emptyIsAll]
  );

  const getAllFilters = (first: boolean, newValue?: string[]) => {
    const filters: IDrawerTreeSelectFilters = {
      search: searchValue.current,
      ...additionalFilters
    };

    if (showLevels) {
      filters.level = levelSelected.current;
    }

    if (formatRender) {
      filters.formats = formatSelected.current;
    }

    if (first) {
      filters.value = value;
    } else {
      filters.value = newValue ? newValue : internalValue;
    }

    return filters;
  };

  const internalLoadData = useCallback(
    async (first: boolean = false, value?: string[]) => {
      const filters = getAllFilters(first, value);

      emptyIsAllRef.current = calcEmptyIsAll(emptyIsAll, filters);

      triggerInputChangeValue(inputRef.current);
      dispatch({
        type: "remoteLoadDataStart",
        payload: value
      });

      const { data, levels, expanded, count } = await loadData(filters);

      mainLevelItems.current = getMainLevelItems(data, levelSelected.current);

      const newState: any = {
        stateTreeData: data
      };

      if (first) {
        prevTreeData.current = data;
      }

      if (count) {
        newState.intenalTreeDataCount = count;
      }

      if (showLevels && levels) {
        newState.internalLevels = levels;
      }

      if (expanded) {
        newState.internalTreeExpandedKeys = expanded;
      }

      if (drawerVisibleRef.current) {
        const s = checkSelectAllStatus(
          value ? value : internalValue,
          !emptyIsAllRef.current
        );

        newState.selectAllState = s.selectAllState;
        newState.internalValue = s.internalValue;
      }

      dispatch({
        type: "remoteLoadDataStop",
        payload: newState
      });

      triggerInputChangeValue(
        inputRef.current,
        remoteSearch ? undefined : searchValue.current
      );
    },
    //eslint-disable-next-line
    [loadData, showLevels, formatRender, remoteSearch, internalValue, value]
  );

  const checkSelectAllStatus = (values, ignoreEmpty: boolean = false) => {
    let checked = true;
    if (!multiple) return;

    if (
      !ignoreEmpty &&
      (!values || !values.length) &&
      emptyIsAll &&
      !searchValue.current
    ) {
      return selectAll();
    }

    if (showCheckedStrategy === "SHOW_PARENT") {
      checked = isAllItemsChecked(values, mainLevelItems.current);
    } else {
      checked = values.length === internalTreeDataCount;
    }
    if (!checked && values.length) {
      return {
        selectAllState: "indeterminate"
      };
    }

    if (checked) {
      return {
        selectAllState: "checked"
      };
    }

    return {
      selectAllState: ""
    };
  };

  const selectAll = () => {
    const state: any = {
      selectAllState: "checked"
    };
    if (mainLevelItems.current) {
      state.internalValue = Array.from(mainLevelItems.current);
    }
    return state;
  };

  const savePrevRefs = () => {
    prevLevel.current = levelSelected.current;
    prevTreeData.current = stateTreeData;
    prevFormatSelected.current = formatSelected.current;
    prevEmptyIsAllRef.current = emptyIsAllRef.current;
  };

  const resetPrevRefs = () => {
    searchValue.current = "";
    prevLevel.current = "1";
    prevTreeData.current = [];
    prevFormatSelected.current = [];
    prevEmptyIsAllRef.current = emptyIsAll;
  };

  const rollbackRefs = () => {
    levelSelected.current = prevLevel.current;
    formatSelected.current = prevFormatSelected.current;

    emptyIsAllRef.current = prevEmptyIsAllRef.current;
    mainLevelItems.current = getMainLevelItems(
      prevTreeData.current,
      levelSelected.current
    );
  };

  const openDrawer = () => {
    drawerVisibleRef.current = true;
    savePrevRefs();

    dispatch({
      type: "openDrawer",
      payload: checkSelectAllStatus(internalValue)
    });
    triggerInputChangeValue(inputRef.current, searchValue.current);
  };

  const closeDrawer = useCallback(() => {
    setTimeout(() => {
      const activeElement = document.activeElement as HTMLElement;
      activeElement.blur();
    }, 100);

    resetPrevRefs();

    drawerVisibleRef.current = false;

    //eslint-disable-next-line
  }, []);

  const isSelectedAll = useMemo(() => {
    return selectAllState === "checked" && emptyIsAllRef.current;
  }, [selectAllState]);

  //  -------- HANDLERS --------

  const handlerDrawerCancel = useCallback(() => {
    const prevValue = !multiple && !value ? [] : value;
    //@ts-ignore
    if (prevValue.length === 0) showAllRef.current = true;

    dispatch({
      type: "drawerCancel",
      payload: {
        stateTreeData: prevTreeData.current,
        internalValue: prevValue
      }
    });
    rollbackRefs();
    closeDrawer();

    setTimeout(() => {
      showAllRef.current = false;
    }, 200);
  }, [closeDrawer, value, multiple, dispatch]);

  const handlerDrawerSubmit = useCallback(() => {
    closeDrawer();
    dispatch({
      type: "drawerSubmit"
    });
    triggerOnChange(internalValue);
  }, [triggerOnChange, closeDrawer, internalValue, dispatch]);

  const handlerDrawerFocus = e => {
    if (internalLoading) return;
    setInputRef(e.target);
    openDrawer();
  };

  const handlerSearchInputChange = useCallback(
    e => {
      searchValue.current = e.target.value;

      if (remoteSearch) {
        internalLoadData();
        return;
      }
      setSearchValue(searchValue.current);

      triggerInputChangeValue(inputRef.current, searchValue.current);
    },
    //eslint-disable-next-line
    [inputRef, internalLoadData]
  );

  const handlerSelectBeforeBlur = useCallback(() => !drawerVisible, [
    drawerVisible
  ]);

  const handleTreeSelect = useCallback(
    (_, node) => {
      dispatch({
        type: "setSelected",
        payload: node
      });
    },
    [dispatch]
  );

  const handleTreeSelectChange = useCallback(
    (value, labels, extra) => {
      let state: any = {};
      if (multiple) {
        state.internalValue = value;
        if (value) {
          const check = checkSelectAllStatus(value, true);
          state = { ...state, ...check };
        }
      } else {
        state.internalValue = extra.checked ? [extra.triggerValue] : [];
      }

      dispatch({
        type: "setState",
        payload: state
      });

      if (!drawerVisible) {
        triggerOnChange(value);
      }
    },
    //eslint-disable-next-line
    [drawerVisible, triggerOnChange, multiple]
  );

  const handlerTreeExpand = useCallback(
    expandedKeys => {
      dispatch({
        type: "internalTreeExpandedKeys",
        payload: expandedKeys
      });
    },
    [dispatch]
  );

  const handleLevelChange = (level: string) => {
    levelSelected.current = level;
    onLevelChange && onLevelChange(level);
    internalLoadData(false, []);
  };

  const handleTreeLoadData = async node => {
    const index = stateTreeData.findIndex(item => item.pId === node.id);
    if (index !== -1) return;
    const data = await loadChildren(node.id, additionalFilters);

    dispatch({
      type: "stateTreeData",
      payload: stateTreeData.concat(data)
    });

    triggerInputChangeValue(inputRef.current, searchValue.current);
  };

  const handleSelectAllChange = e => {
    const checked = e.target.checked;
    let state: any = {};
    if (checked) {
      state = selectAll();
    } else {
      state = {
        selectAllState: "",
        internalValue: []
      };
    }

    dispatch({
      type: "setState",
      payload: state
    });
  };

  // ---- EFFECTS ------

  useEffect(() => {
    dispatch({
      type: "internalValue",
      payload: !multiple && !value ? [] : value
    });
  }, [value, multiple, dispatch]);

  useEffect(() => {
    if (treeData) {
      mainLevelItems.current = getMainLevelItems(
        treeData,
        levelSelected.current
      );
      prevTreeData.current = stateTreeData;
    }

    dispatch({
      type: "setState",
      payload: {
        stateTreeData: treeData,
        internalTreeDataCount: treeDataCount
      }
    });
    //eslint-disable-next-line
  }, [treeData, treeDataCount, dispatch]);

  useEffect(() => {
    dispatch({
      type: "setState",
      payload: {
        internalLoading: loading
      }
    });
  }, [loading, dispatch]);

  useEffect(() => {
    !asyncData && loadData && internalLoadData(true);
    //eslint-disable-next-line
  }, [asyncData, additionalFilters]);

  // -------- RENDERS ---------

  const tagRender = props => {
    if (internalLoading) {
      return (
        <span className="ant-select-selection-placeholder">{loadingText}</span>
      );
    }
    if (isSelectedAll || showAllRef.current) {
      return (
        <span className="ant-select-selection-placeholder">{placeholder}</span>
      );
    }

    return (
      <span className="ant-select-selection-item">
        <Tag
          closable={props.closable}
          onClose={props.onClose}
          className="ant-select-selection-item-content"
        >
          {props.label}
        </Tag>
      </span>
    );
  };

  const format = useMemo(() => {
    if (!formatRender) return null;
    return formatRender({
      onChange: (selected: string[]) => {
        formatSelected.current = selected;
        internalLoadData();
      }
    });
  }, [formatRender, internalLoadData]);

  const dropdownRender = useCallback(
    menu => {
      return (
        <Drawer
          className={clsx({
            "drawer-tree-select-dropdown": true,
            "drawer-tree-select-dropdown-flat-list": isFlatList
          })}
          title={drawerTitle ? drawerTitle : placeholder}
          onClose={handlerDrawerCancel}
          visible={drawerVisible}
          width={
            window.innerWidth < drawerWidth ? window.innerWidth : drawerWidth
          }
          actions={
            <>
              <Button onClick={handlerDrawerCancel}>{cancelText}</Button>
              <Button onClick={handlerDrawerSubmit} type="primary">
                {submitText}
              </Button>
            </>
          }
        >
          {format}
          {isLevelShowed ? (
            <Levels
              onChange={handleLevelChange}
              value={levelSelected.current}
              levels={internalLevels}
            />
          ) : null}
          <SearchInput
            placeholder={drawerSearchPlaceholder}
            value={searchValue.current}
            onChange={handlerSearchInputChange}
            loading={internalLoading}
          />
          <div className="drawer-tree-select-dropdown-toolbar">
            {showSelectAll ? (
              <Checkbox
                onChange={handleSelectAllChange}
                checked={selectAllState === "checked"}
                indeterminate={selectAllState === "indeterminate"}
              >
                {selectAllText}
              </Checkbox>
            ) : null}
          </div>
          {menu}
          <div className="drawer-select-loader-container">
            {internalLoading && (
              <Skeleton
                title={{ width: 300 }}
                paragraph={{ rows: 1 }}
                loading={true}
                active
              />
            )}
          </div>
        </Drawer>
      );
    },
    //eslint-disable-next-line
    [
      drawerVisible,
      searchValue,
      internalLoading,
      internalLevels,
      handlerDrawerCancel,
      handlerDrawerSubmit,
      handlerSearchInputChange
    ]
  );

  const listHeight =
    window.innerHeight -
    204 -
    (formatRender === null ? 0 : 44) -
    (isLevelShowed ? 44 : 0) -
    (showSelectAll ? 34 : 0);

  return (
    <AntTreeSelect
      {...restProps}
      value={internalValue}
      className="drawer-tree-select"
      treeData={stateTreeData}
      open={drawerVisible}
      treeExpandedKeys={internalTreeDefaultExpandedKeys}
      searchValue={searchValue.current ? searchValue.current : ""}
      //@ts-ignore
      dropdownRender={dropdownRender}
      dropdownClassName="drawer-tree-select-dropdown-fake"
      multiple={multiple}
      showSearch={true}
      listHeight={listHeight}
      placeholder={placeholder}
      loading={internalLoading}
      notFoundContent={internalLoading ? loadingText : noDataText}
      showCheckedStrategy={
        searchValue.current ? "SHOW_CHILD" : showCheckedStrategy
      }
      loadData={loadChildren ? handleTreeLoadData : null}
      onBeforeBlur={handlerSelectBeforeBlur}
      onChange={handleTreeSelectChange}
      onFocus={handlerDrawerFocus}
      onTreeExpand={handlerTreeExpand}
      onSelect={handleTreeSelect}
      tagRender={tagRender}
    />
  );
};

DrawerTreeSelect.defaultProps = {
  maxTagCount: 10,
  treeDataCount: 0,
  showLevels: false,
  isFlatList: false,
  drawerTitle: "",
  drawerSearchPlaceholder: "Search",
  drawerWidth: 400,
  cancelText: "Cancel",
  submitText: "Submit",
  loadingText: "Loading",
  noDataText: "No data",
  levelText: "Level %s",
  formatRender: null,
  remoteSearch: false,
  showSelectAll: false,
  selectAllText: "Select all",
  emptyIsAll: false,
  levels: []
};

DrawerTreeSelect.SHOW_ALL = AntTreeSelect.SHOW_ALL;
DrawerTreeSelect.SHOW_CHILD = AntTreeSelect.SHOW_CHILD;
DrawerTreeSelect.SHOW_PARENT = AntTreeSelect.SHOW_PARENT;

export default DrawerTreeSelect;
