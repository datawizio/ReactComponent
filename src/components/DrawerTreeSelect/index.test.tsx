import "jsdom-global/register";
import React from "react";

import DrawerTreeSelect from "./index";
import { shallow } from "enzyme";

import AntTreeSelect from "./antd/AntTreeSelect";

jest.mock("rc-select/es/Selector", () => {});
jest.mock("rc-select/es/SelectTrigger", () => {});

const A = () =>
  React.forwardRef((props, ref) => {
    //@ts-ignore
    return <input ref={ref}>moked</input>;
  });

jest.mock("./antd/AntTreeSelect", () => jest.fn());
//@ts-ignore
AntTreeSelect.mockImplementation(() => A);

describe("DrawerTreeSelect", () => {
  it("DrawerTreeSelect to match snapshot with defaul props", () => {
    const component = shallow(<DrawerTreeSelect />);
    expect(component).toMatchSnapshot();
  });

  it("DrawerTreeSelect to match snapshot with custom props and triger events", () => {
    const mockProps = {
      multiple: true
    };

    const mockChangeEvent = {
      checked: true,
      triggerValue: "value"
    };

    const component = shallow(<DrawerTreeSelect {...mockProps} />);
    component
      .find(".drawer-tree-select")
      .simulate("change", "value", "label", mockChangeEvent);
    component.find(".drawer-tree-select").simulate("select", "", "node");
    component.find(".drawer-tree-select").simulate("beforeBlur");
    component.find(".drawer-tree-select").simulate("treeExpand", "key");

    expect(component).toMatchSnapshot();
  });
});
