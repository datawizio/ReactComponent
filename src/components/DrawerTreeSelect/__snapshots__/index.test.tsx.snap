// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerTreeSelect DrawerTreeSelect to match snapshot with custom props and triger events 1`] = `
<mockConstructor
  className="drawer-tree-select"
  dropdownClassName="drawer-tree-select-dropdown-fake"
  dropdownRender={[Function]}
  listHeight={564}
  loadData={null}
  maxTagCount={10}
  maxTagPlaceholder={[Function]}
  multiple={true}
  notFoundContent="NO_DATA"
  onBeforeBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onSelect={[Function]}
  onTreeExpand={[Function]}
  open={false}
  searchValue=""
  showSearch={true}
  tagRender={[Function]}
  treeCheckStrictly={false}
  treeExpandedKeys="key"
  value="value"
/>
`;

exports[`DrawerTreeSelect DrawerTreeSelect to match snapshot with defaul props 1`] = `
<mockConstructor
  className="drawer-tree-select"
  dropdownClassName="drawer-tree-select-dropdown-fake"
  dropdownRender={[Function]}
  listHeight={564}
  loadData={null}
  maxTagCount={10}
  maxTagPlaceholder={[Function]}
  notFoundContent="NO_DATA"
  onBeforeBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onSelect={[Function]}
  onTreeExpand={[Function]}
  open={false}
  searchValue=""
  showSearch={true}
  tagRender={[Function]}
  treeCheckStrictly={false}
/>
`;
